### ISSUE TITLE
trouble with CloneFunction()

### ISSUE THREAD

**Issue by zhandongdong** [Microsoft/CNTK#1189]
When I try to fine-tune a pre-trained model using new data, I met a problem about CloneFunction().

		featExtNetwork = BS.Network.Load ("model_test")
		featExt = BS.Network.CloneFunction (
              (featExtNetwork.features),    # input node that FE model read data from
              [z3 = featExtNetwork.z3],     # output node in FE model that holds the desired features
              parameters="constant")   # says to freeze that part of the network

               z3 = featExt(myFeatures).z3

I tried to use node "h2" as output node in CloneFunction() , but I can't reach it. It seems that z3 is the only node I can use as output node. 
Does anyone know how to get the node"h2" in model2(x).

    BrainScriptNetworkBuilder = {
        imageShape = 28:28:1                        # image dimensions, 1 channel only
        labelDim = 10                               # number of distinct labels
        featScale = 1/256

        # This model returns multiple nodes as a record, which
        # can be accessed using .x syntax.
		model2(x) = {
            h2 = DenseLayer {100, activation=ReLU} (x)
			drop2 = Dropout(h2)			
            z2 = LinearLayer {labelDim} (drop2)
        }
		
		model(x) = {
            s1 = x * featScale
            h1 = DenseLayer {200, activation=ReLU} (s1)
			drop = Dropout(h1)			
            z = LinearLayer {labelDim} (drop)
			out = model2(z)
        }.out
        

		
        # inputs
        features = Input {imageShape}
        labels = Input {labelDim}

        # apply model to features
        out = model (features)
	 z3 = out.z2
        # loss and error computation
        ce   = CrossEntropyWithSoftmax (labels, z3)
        errs = ClassificationError (labels, z3)

        # declare special nodes
        featureNodes    = (features)
        labelNodes      = (labels)
        criterionNodes  = (ce)
        evaluationNodes = (errs)
        outputNodes     = (z3)

----------------------

**Comment by avader906** [comment]
Simplify then increase complexity if required. Dropout before a linear layer could be questionable. 

If you make layers global in scope, it should work. Please try/simplify to something like this:
```
BrainScriptNetworkBuilder = {
...
    # inputs
    features = Input {imageShape}
    labels = Input {labelDim}

    s1 = features * featScale
    h1 = DenseLayer {200, activation=ReLU} (s1)
    #d1 = Dropout(h1)			#try without first
    z1 = LinearLayer {labelDim} (h1)

    # loss and error computation
    ce   = CrossEntropyWithSoftmax (labels, z1)
    errs = ClassificationError (labels, z1)

    # declare special nodes
    featureNodes    = (features)
    labelNodes      = (labels)
    criterionNodes  = (ce)
    evaluationNodes = (errs)
    outputNodes     = (z1)
...
}
```
then you call recall and use parts of that model with
```
	featExtNetwork = BS.Network.Load ("model_test")
	featExt = BS.Network.CloneFunction (
          featExtNetwork.features,    # input node that FE model read data from
          featExtNetwork.z1,            # output node in FE model that holds the desired features
          parameters="constant")   # says to freeze that part of the network

       z2 = featExt(myFeatures)
```
Try recalling h1 layer (ie specifying the output node as featExtNetwork.h1) instead of the pre-trained linear combination z1 - which could be re-adapted as part of the model extension. Linear combinations do not add/extract new information to the underlying inputs space.

**Comment by zhandongdong** [comment]
@avader906 Thank you for your clear explanation. So it means the node inside function couldn't be reached in CloneFunction. Only global node can be accessed. 


By the way, I found another problem when using CloneFunction(). 

The local variable in function() could be missed and lead to error. 

For example, in function: Splice (inputs, axis=1), default value for axis is 1. But axis is set to 3 in the function below. So it went wrong when I tried to fine-tune the model even by accessing the node in global field.  

out = Splice((branch1x1:branch5x5:branch3x3_3x3:branch_pool), axis=3)
Reason:
It should concatenate the third axis but it tried to concatenate the first axis.  The dimension in third axis is not same which leads to errors. 
  
It's hard to modify the script in this situation, so I don't know how to solve this problem. 


    InceptionBlock1{numOutputChannels1x1, numOutputChannels5x5, numOutputChannels3x3_3x3, numOutputChannelsPool, bnTimeConst} ={
    apply(x) = {
        # 1x1 Convolution
        branch1x1 = ConvBNReLULayer{numOutputChannels1x1, (1:1), (1:1), pad = true, bnTimeConst = bnTimeConst}(x)

        # 5x5 Convolution
        branch5x5 = Sequential(
            ConvBNReLULayer{numOutputChannels5x5[0], (1:1), (1:1), pad = true, bnTimeConst = bnTimeConst} :
            ConvBNReLULayer{numOutputChannels5x5[1], (5:5), (1:1), pad = true, bnTimeConst = bnTimeConst}
        )(x)

        # 3x3 3x3 Convolution
        branch3x3_3x3 = Sequential(
            ConvBNReLULayer{numOutputChannels3x3_3x3[0], (1:1), (1:1), pad = true, bnTimeConst = bnTimeConst} :
            ConvBNReLULayer{numOutputChannels3x3_3x3[1], (3:3), (1:1), pad = true, bnTimeConst = bnTimeConst} :
            ConvBNReLULayer{numOutputChannels3x3_3x3[2], (3:3), (1:1), pad = true, bnTimeConst = bnTimeConst}
        )(x)

        # Average pooling
        branch_pool = Sequential(
            AveragePoolingLayer{(3:3), stride = (1:1), pad = true} :
            ConvBNReLULayer{numOutputChannelsPool, (1:1), (1:1), pad = true, bnTimeConst = bnTimeConst}
        )(x)

        out = Splice((branch1x1:branch5x5:branch3x3_3x3:branch_pool), axis=3)
    }.out }.apply
Error:
Node 'z1.arrayOfFunctions[0].W' (LearnableParameter operation) operation: Tensor shape was inferred as [20 x 27 x 1 x 2048].
Validating --> mixed_1 = RowStack (mixed_1.mixed_1.branch1x1, mixed_1.mixed_1.branch5x5, mixed_1.mixed_1.branch3x3_3x3, mixed_1.mixed_1.branch_pool) : [35 x 35 x 64 x *], [35 x 35 x 64 x *], [35 x 35 x 96 x *], [35 x 35 x 32 x *] -> [140 x 35 x 64 x *] FAILED
……
12/30/2016 08:58:01: EXCEPTION occurred: mixed_1 RowStack operation: Conflicting dimension 2 between mixed_1.mixed_1.branch3x3_3x3 RectifiedLinear operation (96) and other(s) (64)

**Comment by avader906** [comment]
Does this issue relate to this?
```
Problems with node names with . [ and ]

To reference a node in a network that contains . or [ or ], replace those characters 
by _. E.g., if network contains a node called result.z, network.result.z will fail; 
instead say network.result_z.
```
https://github.com/Microsoft/CNTK/wiki/CloneFunction
https://github.com/Microsoft/CNTK/wiki/BS-Model-Editing

**Comment by eldakms** [comment]
Closing the issue due to inactivity, please feel free to reopen if still applies. Thanks!
